src/core/lv2/Lv2Worker.cpp

```diff
@@ -0,0 +1,193 @@
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include <cassert>
+#include <QDebug>
+
+#include "Lv2Worker.h"
+
+#ifdef LMMS_HAVE_LV2
+
+#include "Engine.h"
+
+
+namespace lmms
+{
+
+
+// static wrappers (TODO: avoid those somehow?)
+
+static LV2_Worker_Status staticScheduleWork(LV2_Worker_Schedule_Handle handle,
+	uint32_t size, const void* data)
+{
+	Lv2Worker* worker = static_cast<Lv2Worker*>(handle);
+	return worker->scheduleWork(size, data);
+}
+
+
+
+
+static LV2_Worker_Status
+staticWorkerRespond(LV2_Worker_Respond_Handle handle,
+	uint32_t size, const void* data)
+{
+	Lv2Worker* worker = static_cast<Lv2Worker*>(handle);
+	return worker->respond(size, data);
+}
+
+
+
+
+std::size_t Lv2Worker::bufferSize() const
+{
+	// ardour uses this fixed size for ALSA:
+	return 8192 * 4;
+	// for jack, they use 4 * jack_port_type_get_buffer_size (..., JACK_DEFAULT_MIDI_TYPE)
+	// (possible extension for AudioDevice)
+}
+
+
+
+
+Lv2Worker::Lv2Worker(LV2_Handle handle,
+	const LV2_Worker_Interface* iface,
+	Semaphore& common_work_lock,
+	bool threaded) :
+	m_iface(iface),
+	m_threaded(threaded),
+	m_handle(handle),
+	m_response(bufferSize()),
+	m_requests(bufferSize()),
+	m_responses(bufferSize()),
+	m_requestsReader(m_requests),
+	m_responsesReader(m_responses),
+	m_sem(0),
+	m_workLock(common_work_lock)
+{
+	assert(iface);
+	m_scheduleFeature.handle = static_cast<LV2_Worker_Schedule_Handle>(this);
+	m_scheduleFeature.schedule_work = staticScheduleWork;
+
+	if (threaded) { m_thread = std::thread(&Lv2Worker::workerFunc, this); }
+
+	m_requests.mlock();
+	m_responses.mlock();
+}
+
+
+
+
+Lv2Worker::~Lv2Worker()
+{
+	m_exit = true;
+	if(m_threaded) {
+		m_sem.post();
+		m_thread.join();
+	}
+}
+
+
+
+
+LV2_Worker_Status Lv2Worker::respond(uint32_t size, const void* data)
+{
+	if(!size) { return LV2_WORKER_ERR_UNKNOWN; }
+
+	m_responses.write((const char*)&size, sizeof(size));
+	m_responses.write((const char*)data, size);
+	return LV2_WORKER_SUCCESS;
+}
+
+
+
+
+void Lv2Worker::workerFunc()
+{
+	std::vector<char> buf;
+	uint32_t size;
+	while (true) {
+		m_sem.wait();
+		if (m_exit) break;
+		const std::size_t readSpace = m_requestsReader.read_space();
+		if (readSpace <= sizeof(size)) continue; // (should not happen)
+
+		m_requestsReader.read(sizeof(size)).copy((char*)&size, sizeof(size));
+		Q_ASSERT(readSpace - sizeof(size) >= size);
+		try {
+			buf.resize(size);
+		} catch(...) {
+			qWarning() << "Error: reallocating buffer failed";
+			return;
+		}
+		m_requestsReader.read(size).copy(buf.data(), size);
+
+		m_workLock.wait();
+		m_iface->work(m_handle, staticWorkerRespond, this, size, buf.data());
+		m_workLock.post();
+	}
+}
+
+
+
+
+LV2_Worker_Status Lv2Worker::scheduleWork(uint32_t size, const void *data)
+{
+	if (!size) { return LV2_WORKER_ERR_UNKNOWN; }
+
+	if (m_threaded) {
+		// Schedule a request to be executed by the worker thread
+		m_requests.write((const char*)&size, sizeof(size));
+		m_requests.write((const char*)data, size);
```

Comment:
What if the data doesn't fit in the ringbuffer?

Reply:
The specs don't mention that case, and jalv does not handle this really well. Anyways, I think we have three possibilities:

  1) Crash LMMS
  2) drop the data (send neither size, nor data)
  3) queue the data somehow

(1) seems to aggressive to me (users will lose all changes after last autosave). (3) would require dynamic allocation in the realtime thread (and would make the code look weird). So I guess we should stick with (2)?

In order to have "symmetrical" code, I'd do the same in `Lv2Worker::respond`. Here, dynamic allocation would work, but it sounds like overkill to me...

Reply:
I'm currently switching to (1). It's the safest to just abort, and LMMS will autosave anyways. I'll just use `Q_ASSERT`. Hope that makes sense.

Reply:
Agreed (3) sounds bad. Which of the others is appropriate depends on whether this is a condition one would expect to encounter during normal operation, or whether it's an exceptional condition that indicates something has gone wrong. Since it's entirely permissible for a plugin to call this function with data exceeding the ring buffer size (not the only way for this to fail, but a guaranteed failure), I think the condition is of the former type, and so failure should ideally be handled gracefully.

[The specification states that](https://lv2plug.in/c/html/group__worker.html#a48648b995e558a6cafeae29684c2ad93):
> The host MUST guarantee that [the data] will be eventually passed to `work()` if this function returns `LV2_WORKER_SUCCESS`.

This implies that if the work cannot be delivered, some other value should be returned. Looking at the [`LV2_Worker_Status` enum](https://lv2plug.in/c/html/group__worker.html#ga5f3f4a39cc339d42f0823524940aee9f), there's a `LV2_WORKER_ERR_NO_SPACE` status, described as "failed due to lack of space", which sounds appropriate for this situation. So I would go for (2), dropping the scheduled work and indicating that with the aforementioned return value.

Perhaps a more robust solution could then request that a non-realtime thread allocate a larger ringbuffer, so that the next time a similar piece of work arrives, it can be successfully enqueued. But that's considerably more complex, so I don't think it's worth doing until we (a) find a plugin that delivers such large messages, and (b) confirm that this solution actually benefits that plugin.

I wouldn't use an assertion in this situation, even when implementing (1). Assertions are typically disabled in release mode, which would lead either to the work being silently dropped (if the size can't be written), or to the ringbuffer being corrupted (if the size can be written but not the data). Also, assertions are intended to verify that a condition the programmer believes to be guaranteed to be true, is indeed true. If an assertion fails, it should indicate a bug in the code (or a mistake in the programmer's reasoning), as the program is in a state thought to be impossible to reach. In this case, there is no guarantee or expectation that the ringbuffer will always have space, so an assertion doesn't fit semantically. If you do want to pursue solution (1), I would use a normal if-statement wrapping a call to `std::terminate` or `std::abort`.

Reply:
Agreeing to all. If they even offer a return value for "no space", they probably want us to use it.

I pushed fixups.